lago/sdk_utils.py

```diff
@@ -0,0 +1,74 @@
+import functools
+import inspect
+import wrapt
+
+
+class SDKWrapper(wrapt.ObjectProxy):
+    """A proxy object that exposes only methods which were decorated with
+    :func:`expose` decorator."""
+
+    def __getattr__(self, name):
+        attr = getattr(self.__wrapped__, name)
+        return filter_sdk(attr, name)
+
+    def __dir__(self):
+        orig = super(SDKWrapper, self).__dir__()
+        filtered = []
+        for name in orig:
+            attr = getattr(self.__wrapped__, name)
+            try:
+                filter_sdk(attr, name)
+                filtered.append(name)
+            except AttributeError:
+                pass
+        return filtered
+
+
+class SDKMethod(object):
+    """Metadata to store inside the decorated function"""
+
+    def __init__(self, name):
+        self.name = name
+
+
+def expose(func):
+    """
+    Decorator to be used with :class:`SDKWrapper`. This decorator indicates
+    that the wrapped method or class should be exposed in the proxied object.
+
+    Args:
+        func(types.FunctionType/types.MethodType): function to decorate
+
+    Returns:
+        None
+    """
+
+    @functools.wraps(func)
+    def wrapped(*args, **kwargs):
+        return func(*args, **kwargs)
+
+    if inspect.isclass(func):
+        wrapped._sdkmetaclass = SDKMethod(func.__name__)
+    else:
+        wrapped._sdkmeta = SDKMethod(func.__name__)
+    return wrapped
+
+
+def filter_sdk(attr, name):
+    """
+    Filter SDK attributes
+
+    Args:
+        attr(attribute): Attribute as returned by :func:`getattr`.
+        name(str): Attribute name.
+
+    Returns:
+        `attr` if passed.
+    """
+
+    if name.startswith('_'):
+        return attr
```

Comment:
I think that a "good" UX is when you import the SDK you only have public methods you are supposed to use and have documentation. If you "look" at the underscore methods, you know its on your own risk. You can also access ``SDK._prefix`` and get the "normal" object directly, but again that is internal and can change.

 As to why (still) use the proxy - we have many other "public" methods in prefix that are either:
1. Undocumented.
2. Do not make sense for SDK use(but do for CLI)(like ``save``, ``initialize``, ``resolve_parent``, ``fetch_url``, ``virt_conf_from_stream``, etc).

With the proxy - we can gradually change the internal implementation without breaking the SDK - like moving the 'do_init' to the SDK, etc.

I can filter the '_' methods also, its not a big issue.